File: jogl-samples/src/tests/gl_300/Gl_300_test_alpha.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package tests.gl_300;

import com.jogamp.opengl.GL;
import static com.jogamp.opengl.GL.GL_ARRAY_BUFFER;
import static com.jogamp.opengl.GL.GL_COLOR_BUFFER_BIT;
import static com.jogamp.opengl.GL.GL_FLOAT;
import static com.jogamp.opengl.GL.GL_GREATER;
import static com.jogamp.opengl.GL.GL_NEAREST;
import static com.jogamp.opengl.GL.GL_STATIC_DRAW;
import static com.jogamp.opengl.GL.GL_TEXTURE0;
import static com.jogamp.opengl.GL.GL_TEXTURE_2D;
import static com.jogamp.opengl.GL.GL_TEXTURE_MAG_FILTER;
import static com.jogamp.opengl.GL.GL_TEXTURE_MIN_FILTER;
import static com.jogamp.opengl.GL.GL_TRIANGLES;
import com.jogamp.opengl.GL2ES1;
import static com.jogamp.opengl.GL2ES1.GL_ALPHA_TEST;
import static com.jogamp.opengl.GL2ES2.GL_FRAGMENT_SHADER;
import static com.jogamp.opengl.GL2ES2.GL_MAX_VARYING_VECTORS;
import static com.jogamp.opengl.GL2ES2.GL_VERTEX_SHADER;
import static com.jogamp.opengl.GL2ES3.GL_MAX_VARYING_COMPONENTS;
import com.jogamp.opengl.GL3;
import com.jogamp.opengl.math.FloatUtil;
import com.jogamp.opengl.util.GLBuffers;
import com.jogamp.opengl.util.glsl.ShaderCode;
import com.jogamp.opengl.util.glsl.ShaderProgram;
import framework.BufferUtils;
import framework.Profile;
import framework.Semantic;
import framework.Test;
import java.io.IOException;
import java.nio.FloatBuffer;
import java.util.logging.Level;
import java.util.logging.Logger;
import jgli.Texture2d;

/**
 *
 * @author GBarbieri
 */
public class Gl_300_test_alpha extends Test {

    public static void main(String[] args) {
        Gl_300_test_alpha gl_300_test_alpha = new Gl_300_test_alpha();
    }

    public Gl_300_test_alpha() {
        super("gl-300-test-alpha", Profile.COMPATIBILITY, 3, 2);
    }

    private final String SHADERS_SOURCE = "image-2d";
    private final String SHADERS_ROOT = "src/data/gl_300";
    private final String TEXTURE_DIFFUSE = "kueken7_rgba8_srgb.dds";

    public static class Vertex {

        public float[] position;
        public float[] texCoord;
        public static final int SIZEOF = 2 * 2 * Float.BYTES;

        public Vertex(float[] position, float[] texCoord) {
            this.position = position;
            this.texCoord = texCoord;
        }

        public float[] toFloatArray() {
            return new float[]{position[0], position[1], texCoord[0], texCoord[1]};
        }
    }

    // With DDS textures, v texture coordinate are reversed, from top to bottom
    private int vertexCount = 6;
    private int vertexSize = vertexCount * Vertex.SIZEOF;
    private Vertex[] vertexData = {
        new Vertex(new float[]{-1.0f, -1.0f}, new float[]{0.0f, 1.0f}),
        new Vertex(new float[]{+1.0f, -1.0f}, new float[]{1.0f, 1.0f}),
        new Vertex(new float[]{+1.0f, +1.0f}, new float[]{1.0f, 0.0f}),
        new Vertex(new float[]{+1.0f, +1.0f}, new float[]{1.0f, 0.0f}),
        new Vertex(new float[]{-1.0f, +1.0f}, new float[]{0.0f, 0.0f}),
        new Vertex(new float[]{-1.0f, -1.0f}, new float[]{0.0f, 1.0f})};

    private int[] vertexArrayName = new int[1], bufferName = new int[1], texture2dName = new int[1];
    private int programName, uniformMvp, uniformDiffuse;
    private float[] projection = new float[16], model = new float[16], mvp = new float[16];

    @Override
    protected boolean begin(GL gl) {

        GL3 gl3 = (GL3) gl;

        boolean validated = true;

        int[] maxVaryingOutputComp = {0};
        gl3.glGetIntegerv(GL_MAX_VARYING_COMPONENTS, maxVaryingOutputComp, 0);
        int[] maxVaryingOutputVec = {0};
        gl3.glGetIntegerv(GL_MAX_VARYING_VECTORS, maxVaryingOutputVec, 0);

        if (validated) {
            validated = initTest(gl3);
        }
        if (validated) {
            validated = initProgram(gl3);
        }
        if (validated) {
            validated = initBuffer(gl3);
        }
        if (validated) {
            validated = initVertexArray(gl3);
        }
        if (validated) {
            validated = initTexture(gl3);
        }

        return validated && checkError(gl3, "begin");
    }

    protected boolean initTest(GL3 gl3) {

        gl3.glEnable(GL_ALPHA_TEST);
        ((GL2ES1) gl3).glAlphaFunc(GL_GREATER, 0.2f);

        //To test alpha blending:
//        gl3.glEnable(GL_BLEND);
//        gl3.glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
        return checkError(gl3, "initTest");
    }

    private boolean initProgram(GL3 gl3) {

        boolean validated = true;

        if (validated) {

            ShaderCode vertShaderCode = ShaderCode.create(gl3, GL_VERTEX_SHADER,
                    this.getClass(), SHADERS_ROOT, null, SHADERS_SOURCE, "vert", null, true);
            ShaderCode fragShaderCode = ShaderCode.create(gl3, GL_FRAGMENT_SHADER,
                    this.getClass(), SHADERS_ROOT, null, SHADERS_SOURCE, "frag", null, true);

            ShaderProgram shaderProgram = new ShaderProgram();
            shaderProgram.add(vertShaderCode);
            shaderProgram.add(fragShaderCode);

            shaderProgram.init(gl3);

            programName = shaderProgram.program();

            gl3.glBindAttribLocation(programName, Semantic.Attr.POSITION, "position");
            gl3.glBindAttribLocation(programName, Semantic.Attr.TEXCOORD, "texCoord");

            shaderProgram.link(gl3, System.out);
        }
        if (validated) {

            uniformMvp = gl3.glGetUniformLocation(programName, "mvp");
            uniformDiffuse = gl3.glGetUniformLocation(programName, "diffuse");
        }

        return validated & checkError(gl3, "initProgram");
    }

    private boolean initBuffer(GL3 gl3) {

        gl3.glGenBuffers(1, bufferName, 0);

        gl3.glBindBuffer(GL_ARRAY_BUFFER, bufferName[0]);
        FloatBuffer vertexBuffer = GLBuffers.newDirectFloatBuffer(vertexCount * 4);
        for (Vertex vertex : vertexData) {
            vertexBuffer.put(vertex.toFloatArray());
        }
        gl3.glBufferData(GL_ARRAY_BUFFER, vertexSize, vertexBuffer.rewind(), GL_STATIC_DRAW);
        BufferUtils.destroyDirectBuffer(vertexBuffer);
        gl3.glBindBuffer(GL_ARRAY_BUFFER, 0);

        return checkError(gl3, "initBuffer");
    }

    private boolean initTexture(GL3 gl3) {

        try {
            gl3.glGenTextures(1, texture2dName, 0);

            gl3.glActiveTexture(GL_TEXTURE0);
            gl3.glBindTexture(GL_TEXTURE_2D, texture2dName[0]);
            gl3.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
            gl3.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);

            jgli.Texture2d texture = new Texture2d(jgli.Load.load(TEXTURE_ROOT + "/" + TEXTURE_DIFFUSE));
            jgli.Gl.Format format = jgli.Gl.instance.translate(texture.format());
            for (int level = 0; level < texture.levels(); ++level) {
                gl3.glTexImage2D(GL_TEXTURE_2D, level,
                        format.internal.value,
                        texture.dimensions(level)[0], texture.dimensions(level)[1],
                        0,
                        format.external.value, format.type.value,
                        texture.data(0, 0, level));
            }

        } catch (IOException ex) {
            Logger.getLogger(Gl_300_test_alpha.class.getName()).log(Level.SEVERE, null, ex);
        }
        return checkError(gl3, "initTexture");
    }

    private boolean initVertexArray(GL3 gl3) {

        gl3.glGenVertexArrays(1, vertexArrayName, 0);
        gl3.glBindVertexArray(vertexArrayName[0]);
        {
            gl3.glBindBuffer(GL_ARRAY_BUFFER, bufferName[0]);
            gl3.glVertexAttribPointer(Semantic.Attr.POSITION, 2, GL_FLOAT, false, Vertex.SIZEOF, 0);
            gl3.glVertexAttribPointer(Semantic.Attr.TEXCOORD, 2, GL_FLOAT, false, Vertex.SIZEOF, 2 * Float.BYTES);
            gl3.glBindBuffer(GL_ARRAY_BUFFER, 0);

            gl3.glEnableVertexAttribArray(Semantic.Attr.POSITION);
            gl3.glEnableVertexAttribArray(Semantic.Attr.TEXCOORD);
        }
        gl3.glBindVertexArray(0);

        return checkError(gl3, "initVertexArray");
    }

    @Override
    protected boolean render(GL gl) {

        GL3 gl3 = (GL3) gl;

        FloatUtil.makePerspective(projection, 0, true, (float) Math.PI * 0.25f,
                (float) windowSize.x / windowSize.y, 0.1f, 100.0f);
        FloatUtil.makeIdentity(model);
        FloatUtil.multMatrix(projection, view(), mvp);
        FloatUtil.multMatrix(mvp, model);

        gl3.glViewport(0, 0, windowSize.x, windowSize.y);
        gl3.glClearColor(1.0f, 0.5f, 0.0f, 1.0f);
        gl3.glClear(GL_COLOR_BUFFER_BIT);

        gl3.glUseProgram(programName);
        gl3.glUniform1i(uniformDiffuse, 0);
        gl3.glUniformMatrix4fv(uniformMvp, 1, false, mvp, 0);

        gl3.glActiveTexture(GL_TEXTURE0);
        gl3.glBindTexture(GL_TEXTURE_2D, texture2dName[0]);

        gl3.glBindVertexArray(vertexArrayName[0]);

        gl3.glDrawArrays(GL_TRIANGLES, 0, vertexCount);

        return true;
    }

    @Override
    protected boolean end(GL gl) {

        GL3 gl3 = (GL3) gl;

        gl3.glDeleteBuffers(1, bufferName, 0);
        gl3.glDeleteProgram(programName);
        gl3.glDeleteTextures(1, texture2dName, 0);
        gl3.glDeleteVertexArrays(1, vertexArrayName, 0);

        return true;
    }
}

File: src/main/kotlin/oglSamples/tests/gl300/gl-300-test-alpha.kt
package oglSamples.tests.gl300

import gli_.Texture2d
import gli_.gl
import gli_.gli
import glm_.glm
import glm_.mat4x4.Mat4
import glm_.vec2.Vec2
import gln.BufferTarget.Companion.ARRAY
import gln.TextureTarget.Companion._2D
import gln.cap.Caps
import gln.draw.glDrawArrays
import gln.glViewport
import gln.glf.glf
import gln.objects.GlProgram
import gln.objects.GlTexture
import gln.texture.TexFilter
import gln.uniform.glUniform
import gln.vertexArray.glEnableVertexAttribArray
import gln.vertexArray.glVertexAttribPointer
import oglSamples.*
import oglSamples.framework.Framework
import oglSamples.framework.semantic
import org.lwjgl.opengl.GL11.GL_ALPHA_TEST
import org.lwjgl.opengl.GL11.glAlphaFunc
import org.lwjgl.opengl.GL11C.*
import org.lwjgl.opengl.GL30C.GL_MAX_VARYING_COMPONENTS
import org.lwjgl.opengl.GL41C.GL_MAX_VARYING_VECTORS

fun main() {
    gl_300_test_alpha()()
}

class gl_300_test_alpha : Framework("gl-300-test-alpha", Caps.Profile.COMPATIBILITY, 3, 0) {

    val SHADER_SOURCE = "gl-300/image-2d"
    val TEXTURE_DIFFUSE = "kueken7_rgba8_srgb.dds"

    class Vertex(val position: Vec2, val texCoord: Vec2) {
        companion object {
            val size = Vec2.size * 2
        }
    }

    // With DDS textures, v texture coordinate are reversed, from top to bottom
    val vertexCount = 6
    val vertexSize = vertexCount * Vertex.size
    val vertexData = vertex_v2v2_buffer_of(
            Vertex_v2v2(Vec2(-1f, -1f), Vec2(0f, 1f)),
            Vertex_v2v2(Vec2(+1f, -1f), Vec2(1f, 1f)),
            Vertex_v2v2(Vec2(+1f, +1f), Vec2(1f, 0f)),
            Vertex_v2v2(Vec2(+1f, +1f), Vec2(1f, 0f)),
            Vertex_v2v2(Vec2(-1f, +1f), Vec2(0f, 0f)),
            Vertex_v2v2(Vec2(-1f, -1f), Vec2(0f, 1f)))

    var vertexArray = GlVertexArray()
    var program = GlProgram.NULL
    var buffer = GlBuffer()
    var texture2D = GlTexture()
    var uniformMVP = -1
    var uniformDiffuse = -1

    override fun begin(): Boolean {

        var validated = true

        val maxVaryingOutputComp = glGetInteger(GL_MAX_VARYING_COMPONENTS)
        val maxVaryingOutputVec = glGetInteger(GL_MAX_VARYING_VECTORS)

        if (validated)
            validated = initTest()
        if (validated)
            validated = initProgram()
        if (validated)
            validated = initBuffer()
        if (validated)
            validated = initVertexArray()
        if (validated)
            validated = initTexture()

        return validated && checkError("begin")
    }

    fun initTest(): Boolean {
        glEnable(GL_ALPHA_TEST)
        glAlphaFunc(GL_GREATER, 0.2f)

        //To framework alpha blending:
        //glEnable(GL_BLEND);
        //glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

        return checkError("initTest")
    }

    fun initProgram(): Boolean {

        var validated = true

        try {
            program = GlProgram.initFromPath("$SHADER_SOURCE.vert", "$SHADER_SOURCE.frag") {
                "Position".attrib = semantic.attr.POSITION
                "Texcoord".attrib = semantic.attr.TEXCOORD
            }
        } catch (exc: Exception) {
            validated = false
        }

        if (validated) {
            uniformMVP = program getUniformLocation "MVP"
            uniformDiffuse = program getUniformLocation "Diffuse"
        }

        return validated && checkError("initProgram")
    }

    fun initBuffer(): Boolean {

        buffer = GlBuffer.gen().bound(ARRAY) {
            data(vertexData.data)
        }

        return checkError("initBuffer")
    }

    fun initTexture(): Boolean {

        texture2D = GlTexture.gen().bound(_2D,0) {

            minMagFilter = TexFilter.NEAREST

            gli.gl.profile = gl.Profile.GL32
            val dds = Texture2d(gli.loadDds(ClassLoader.getSystemResource(TEXTURE_DIFFUSE).toURI()))
            val format = gli.gl.translate(dds.format, dds.swizzles)
            for (level in 0 until dds.levels())
                image(level,
                        format.internal,
                        dds[level].extent(),
                        format.external, format.type,
                        dds[level].data()!!)
        }
        return checkError("initTexture")
    }

    fun initVertexArray(): Boolean {

        vertexArray = GlVertexArray.gen().bound {
            buffer.bound(ARRAY) {
                glVertexAttribPointer(glf.pos2_tc2)
            }
            glEnableVertexAttribArray(glf.pos2_tc2)
        }

        return checkError("initVertexArray")
    }

    override fun render(): Boolean {

        val projection = glm.perspective(glm.πf * 0.25f, windowSize.aspect, 0.1f, 100f)
        val model = Mat4(1f)
        val mvp = projection * view * model

        glViewport(windowSize)
        glClearColor(1f, 0.5f, 0f, 1f)
        glClear(GL_COLOR_BUFFER_BIT)

        program.use()
        glUniform(uniformDiffuse, 0)
        glUniform(uniformMVP, mvp)

        texture2D.bind(_2D,0)

        vertexArray.bind()

        glDrawArrays(vertexCount)

        return true
    }

    override fun end(): Boolean {

        buffer.delete()
        program.delete()
        texture2D.delete()
        vertexArray.delete()

        return true
    }
}